elliptic-curve/src/group_digest.rs

```diff
@@ -0,0 +1,86 @@
+use crate::hash2field::{hash_to_field, ExpandMsg, FromOkm};
+use crate::isogeny::Isogeny;
+use crate::osswu::OsswuMap;
+use ff::Field;
+use group::Group;
+
+/// Convert a set of field elements into a point
+pub trait ConvertToPoint {
+    /// The field element that represents a point
+    type FieldElement: Field;
+    /// The result of converting two field elements to a Point
+    type Output;
+
+    /// Converts two coordinates into a point
+    fn convert_to_point(x: Self::FieldElement, y: Self::FieldElement) -> Self::Output;
+}
+
+/// Method for points to be cleared by the curve cofactor
+pub trait ClearCofactor {
+    /// Multiply this point by the curve cofactor to ensure
+    /// correct subgroup membership
+    fn clear_cofactor(&mut self);
+}
+
+/// Adds hashing arbitrary byte sequences to a valid group element
+pub trait GroupDigest {
+    /// The field element representation for a group value with multiple elements
+    type FieldElement: FromOkm + OsswuMap + Isogeny;
+    /// The resulting group element
+    type Output: Group
+        + ClearCofactor
+        + ConvertToPoint<FieldElement = Self::FieldElement, Output = Self::Output>;
+
+    /// Computes the hash to curve routine according to
+    /// <https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html>
+    /// which says
+    /// Uniform encoding from byte strings to points in G.
+    /// That is, the distribution of its output is statistically close
+    /// to uniform in G.
+    /// This function is suitable for most applications requiring a random
+    /// oracle returning points in G assuming a cryptographically secure
+    /// hash function is used.
+    ///
+    /// Example using a fixed size hash function
+    ///
+    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXmd<sha2::Sha256>>(b"test data", b"CURVE_XMD:SHA-256_SSWU_RO_");
+    ///
+    ///
+    /// Example using an extendable output function
+    ///
+    /// let pt = ProjectivePoint::hash_from_bytes::<hash2field::ExpandMsgXof<sha3::Shake256>>(b"test data", b"CURVE_XOF:SHAKE-256_SSWU_RO_");
+    ///
+    fn hash_from_bytes<X: ExpandMsg, M: AsRef<[u8]>>(msg: M, dst: &'static [u8]) -> Self::Output {
```

Comment:
```suggestion
    fn hash_from_bytes<X: ExpandMsg, M: AsRef<[u8]>>(msg: M, dst: &[u8]) -> Self::Output {
```
Is there a reason this is `'static`?

Reply:
Yes. HashToCurve requires handling a DST longer than 255 chars. See `Domain`. Having implemented this multiple times across various languages and libraries, 99% (I've never seen a dynamic one yet but there must be one somewhere) of the time the domain is static or known and not composed at run time. To be `no-std` compliant I can't use `Vec`. That leaves a couple of options: use `[u8; 255]` or `GenericArray<u8, U255>`, use `&'a [u8]` or `&'static [u8]`. The first option allocates more than enough space to handle any DST < 255 but overly so since they are usually < 32. This also requires copying the DST into the other array which is not really necessary. That leaves the other two. The first `&'a [u8]` requires adding the lifetime to all traits up the chain and more complex work to make Rust happy. This is undesirable and complicated. Thus the use of `&'static [u8]` which is not that strict of a requirement and keeps the API clean.

Reply:
I actually thought that `&'a [u8]` was a perfectly fine solution here. This can be improved in a follow-up PR, not really a big issue.